src/Haskell/Ide/Engine/Plugin/HieExtras.hs

```diff
@@ -100,91 +98,100 @@ instance ModuleCache NameMapData where
-
-              processSig :: T.Text
-                         -> Sig GM.GhcPs
-                         -> [(J.SymbolKind, Located T.Text, Maybe T.Text)]
-              processSig cnt (ClassOpSig False names _) =
-                map (\n ->(J.SkMethod,s n, Just cnt)) names
-              processSig _ _ = []
-
-              processCon :: T.Text
-                         -> ConDecl GM.GhcPs
-                         -> [(J.SymbolKind, Located T.Text, Maybe T.Text)]
-              processCon cnt ConDeclGADT { con_names = names } =
-                map (\n -> (J.SkConstructor, s n, Just cnt)) names
-              processCon cnt ConDeclH98 { con_name = name, con_details = dets } =
-                (J.SkConstructor, sn, Just cnt) : xs
-                where
-                  sn = s name
-                  xs = case dets of
-                    RecCon (L _ rs) -> concatMap (map (f . rdrNameFieldOcc . unLoc)
-                                                 . cd_fld_names
-                                                 . unLoc) rs
-                                         where f ln = (J.SkField, s ln, Just (unLoc sn))
-                    _ -> []
-
-              goImport :: ImportDecl GM.GhcPs -> [(J.SymbolKind, Located T.Text, Maybe T.Text)]
-              goImport ImportDecl { ideclName = lmn, ideclAs = as, ideclHiding = meis } = a ++ xs
-                where
-                  im = (J.SkModule, lsmn, Nothing)
-                  lsmn = s lmn
-                  smn = unLoc lsmn
-                  a = case as of
-                            Just a' -> [(J.SkNamespace, lsmn, Just $ showName a')]
-                            Nothing -> [im]
-                  xs = case meis of
-                         Just (False, eis) -> concatMap (f . unLoc) (unLoc eis)
-                         _ -> []
-                  f (IEVar n) = pure (J.SkFunction, s n, Just smn)
-                  f (IEThingAbs n) = pure (J.SkClass, s n, Just smn)
-                  f (IEThingAll n) = pure (J.SkClass, s n, Just smn)
-                  f (IEThingWith n _ vars fields) =
-                    let sn = s n in
-                    (J.SkClass, sn, Just smn) :
-                         map (\n' -> (J.SkFunction, s n', Just (unLoc sn))) vars
-                      ++ map (\f' -> (J.SkField   , s f', Just (unLoc sn))) fields
-                  f _ = []
-
-              declsToSymbolInf :: (J.SymbolKind, Located T.Text, Maybe T.Text)
-                               -> IdeM (Either T.Text J.SymbolInformation)
-              declsToSymbolInf (kind, L l nameText, cnt) = do
-                eloc <- srcSpan2Loc rfm l
-                case eloc of
-                  Left x -> return $ Left x
-                  Right loc -> return $ Right $ J.SymbolInformation nameText kind loc cnt
-          symInfs <- mapM declsToSymbolInf (imps ++ decls)
-          return $ IdeResponseOk $ rights symInfs
+getSymbols :: Uri -> (IdeResponse [J.SymbolInformation] -> IdeM ()) -> IdeM ()
```

Comment:
@wz1000 I'm not the biggest fan of this callback part. Is there a better place/way to do this?

Reply:
working on something